pywps/inout/basic.py

```diff
@@ -254,58 +295,100 @@ def clone(self):
-        >>> isinstance(h, DataHandler)
-        True
-
-        Note that trying to set another attribute (e.g. `h.file = 'a.txt'`) will raise an AttributeError.
-        """
-        for cls in (FileHandler, DataHandler, StreamHandler, UrlHandler):
-            def fset(s, value, kls=cls):
-                """Assign the handler class and set the value to the attribute.
+    @property
+    def base64(self):
+        """Return raw data
+        WARNING: may be bytes or str"""
+        return self._iohandler.base64
+
+    @property
+    def file(self):
+        """Return a file name"""
+        return self._iohandler.file
+
+    @file.setter
+    def file(self, value):
+        self._iohandler = FileHandler(value, self)
+        self._check_valid()
+
+    @property
+    def data(self):
+        """Return raw data
+        WARNING: may be bytes or str"""
+        return self._iohandler.data
+
+    @data.setter
+    def data(self, value):
+        self._iohandler = DataHandler(value, self)
+        self._check_valid()
 
-                This function will only be called once. The next `fset` will
-                use the subclass' property.
-                """
-                # Add cls methods to this instance.
-                extend_instance(s, kls)
+    @property
+    def stream(self):
+        """Return stream of data
+        WARNING: may be FileIO or StringIO"""
+        return self._iohandler.stream
 
-                # Set the attribute value through the associated cls property.
-                setattr(s, kls.prop, value)
+    @stream.setter
+    def stream(self, value):
+        self._iohandler = StreamHandler(value, self)
+        self._check_valid()
 
-            setattr(IOHandler, cls.prop, property(fget=lambda x: None, fset=fset))
+    @property
+    def url(self):
+        """Return the url of data"""
+        return self._iohandler.url
+
+    @url.setter
+    def url(self, value):
+        self._iohandler = UrlHandler(value, self)
+        self._check_valid()
+
+    # FIXME: post_data is only related to url, this should be initialize with url setter
+    @property
+    def post_data(self):
+        return self._iohandler.post_data
 
+    # Will raise an arror if used on invalid object
+    @post_data.setter
+    def post_data(self, value):
+        self._iohandler.post_data = value
 
-class FileHandler(IOHandler):
+    @property
+    def prop(self):
+        return self._iohandler.prop
+
+    @prop.setter
+    def prop(self, value):
+        LOGGER.warning("Deprecated use of IOHandler.prop, prefer self.value = self.value")
+        if value == "file":
+            self._iohandler = FileHandler(self._iohandler.file, self)
+        elif value == "data":
+            self._iohandler = DataHandler(self._iohandler.data, self)
+        elif value == "stream":
+            self._iohandler = StreamHandler(self._iohandler.stream, self)
+        elif value == "url":
+            self._iohandler = UrlHandler(self._iohandler.url, self)
```

Comment:
Do we really need this ? Removing it does not create any test failure. I don't think this was supported previously.

Reply:
Hello huard,

At beginning I did not implement the prop setter but the test : test_inout.py:test_complex_input_stream need it.

Otherwise, I agree that we should remove it.